### PR TITLE
More refactors and cleanups

### PR DESCRIPTION
**Pull request by aweebs**
- Refactor the store-sync code into smaller functions which can be composed better, even if only a few are used
- Adds much more explicit checks to satisfy static type checking for WebSocketMessage handling
- Added a parser for verifying the external config is the expected shape
- Add an eslint rule to ban `@ts-` directives, and replaces them with casts. This is minimally safer but prevents multiple unexpected errors from being hidden on a line.
- Upgrades typescript dependency to version 5
- A bunch of minor typefixes and teaks. This resolves almost every non-vuetify-bad-typing error.

Recommend reviewing commit by commit and with whitespace visualization disabled.

### CODE REVIEW

File: frontend/src/utils/store-sync.ts [hunk 1]
```diff
@@ -1,161 +1,206 @@
...
-): Promise<T | DisplayPreferencesDto> {
+export async function fetchDisplayPreferences(
+  displayPreferencesId: string
+): Promise<DisplayPreferencesDto> {
   const remote = useRemote();
 
+  if (!remote.auth.currentUserId) {
+    throw new Error(USER_UNAUTHENTICATED_MESSAGE);
+  }
+
```
Comment (aweebs) [inline frontend/src/utils/store-sync.ts:40]: This check isn't really done anywhere else, we usually just default the currentUserId if it's null, so I'll likely just do that instead of this error check everywhere.

File: frontend/src/utils/external-config.ts [hunk 2]
```diff
@@ -1,22 +1,57 @@
...
 /**
  * Fetch configuration at runtime from the config.json file
  * We use destr for serialization as it has better support for JS primitives.
  */
 export async function getJSONConfig(): Promise<ExternalJSONConfig> {
   if (isNil(externalConfig)) {
-    externalConfig = destr(
-      JSON.stringify(await (await fetch('/config.json')).json())
-    ) as ExternalJSONConfig;
+    const loadedConfig = await (await fetch('/config.json')).json();
```
Comment (aweebs) [inline frontend/src/utils/external-config.ts:50]: `fetch` would have already tried to parse the response with the regular JSON parser and not `destr`, so this second stringify and parsing was pointless.

File: frontend/src/utils/playback-profiles/helpers/codec-profiles.ts [hunk 3]
```diff
@@ -26,7 +26,17 @@ import {
...
+    isTizen() &&
+    'webapis' in window &&
+    typeof window.webapis === 'object' &&
+    window.webapis &&
+    'productinfo' in window.webapis &&
+    typeof window.webapis.productinfo === 'object' &&
+    window.webapis.productinfo &&
+    'isUdPanelSupported' in window.webapis.productinfo &&
+    typeof window.webapis.productinfo.isUdPanelSupported === 'function'
+  ) {
```
Comment (ferferga) [inline frontend/src/utils/playback-profiles/helpers/codec-profiles.ts:39]: FYI, our idea is to move completely out from our custom profile detections and use MediaCapabilities API, so this file is indeed deprecated.

You can coordinate with @ThibaultNocchi, because it will be a huge step forward better playback and this is the major legacy thing we have in this client as of today.

File: frontend/src/utils/store-sync.ts [hunk 4]
```diff
@@ -1,161 +1,206 @@
 import { DisplayPreferencesDto } from '@jellyfin/sdk/lib/generated-client';
 import { getDisplayPreferencesApi } from '@jellyfin/sdk/lib/utils/api/display-preferences-api';
 import destr from 'destr';
-import { isNil } from 'lodash-es';
 import { usei18n, useRemote, useSnackbar } from '@/composables';
 import { taskManagerStore } from '@/store';
 
+const CLIENT = 'vue';
+
+const USER_UNAUTHENTICATED_MESSAGE = 'Expected user to be authenticated';
```
Comment (ferferga) [inline frontend/src/utils/store-sync.ts:9]:
```suggestion
const CLIENT = 'vue';
const USER_UNAUTHENTICATED_MESSAGE = 'Expected user to be authenticated';
```

File: frontend/src/utils/store-sync.ts [hunk 5]
```diff
@@ -1,161 +1,206 @@
...
- * Pushes a new displayPreferences payload to server
+ * Uses the keys on `defaults` to extract values from the Server's CustomPrefs
+ * and de-serializes them to a value.
+ * All keys needed for default should exist on the defaults parameter.
+ * Warning: No runtime checking is performed and de-serialized data could vary in shape from what is expected.
  */
-export async function pushSettingsToServer(
-  key: string,
-  prefs: DisplayPreferencesDto
+export async function fetchDefaultedCustomPrefs<T extends object>(
```
Comment (ferferga) [inline frontend/src/utils/store-sync.ts:122]: It's not clear to me yet why the function rename is needed

**Review by ferferga — CHANGES_REQUESTED** [review]

File: frontend/src/components/Buttons/Playback/PlaybackSettingsButton.vue [hunk 6]
```diff
@@ -8,8 +8,7 @@
       <tooltip-button
```
[the remark this answers is not in the document]
Comment (aweebs) [reply frontend/src/components/Buttons/Playback/PlaybackSettingsButton.vue:8]: Agree with this since the component is really basic and doesn't have any real special/custom implementation. If we're all aligned on that being the way forward I can add it to this PR or do it separately. Will wait until we have consensus.

File: frontend/src/components/Layout/Images/__tests__/BlurhashCanvas.spec.ts [hunk 7]
```diff
@@ -14,35 +14,26 @@ beforeEach(() => {
 
 describe('component: BlurhashCanvas', () => {
   it('has a hash property of type String', () => {
-    // @ts-expect-error - props are not defined in the Vue type.
```
[the remark this answers is not in the document]
Comment (aweebs) [reply frontend/src/components/Layout/Images/__tests__/BlurhashCanvas.spec.ts:17]: Yeah, I did it more for completeness :D

Agree that we should have a `tests` directory alongside `src`. Do you want that part of this PR?

File: frontend/src/pages/index.vue [hunk 8]
```diff
@@ -62,29 +53,13 @@ onMounted(async () => {
...
-      'latestmedia'
-    );
-  }
+  const homeSectionKeys = [
+    'librarytiles',
+    'resume',
+    'resumeaudio',
+    'upnext',
+    'latestmedia'
+  ];
```
[the remark this answers is not in the document]
Comment (aweebs) [reply frontend/src/pages/index.vue:62]: It does use some reactive values from userLibraries on line 88

File: frontend/src/utils/playback-profiles/helpers/codec-profiles.ts (hunk 3 above)
Comment (aweebs) [reply frontend/src/utils/playback-profiles/helpers/codec-profiles.ts:39]: Do you want to coordinate removing it in this PR or for future work?

File: frontend/src/utils/external-config.ts [hunk 9]
```diff
@@ -1,22 +1,57 @@
...
+
+  if (
+    !('defaultServerURLs' in config) ||
+    !Array.isArray(config.defaultServerURLs)
+  ) {
+    throw new Error('Expected defaultServerURLS array');
+  }
+
+  if (
+    config.defaultServerURLs.some(
```
[the remark this answers is not in the document]
Comment (aweebs) [reply frontend/src/utils/external-config.ts:28]: Ah, this should be some but the inner check is wrong, and should check if any of them _are not_ strings. Good catch!

File: frontend/src/store/clientSettings.ts [hunk 10]
```diff
@@ -143,7 +143,7 @@ class ClientSettingsStore {
       async () => {
         if (remote.auth.currentUser) {
           try {
-            const data = await fetchSettingsFromServer<ClientSettingsState>(
+            const data = await fetchDefaultedCustomPrefs(
```
[the remark this answers is not in the document]
Comment (aweebs) [reply frontend/src/store/clientSettings.ts:146]: The rename was meant to be more clear about what it does (obviously that didn't work 😅 ).
Overall functionality is the same to how it worked in practice previously. Given a default object, it takes the properties from the default object and checks if they exist in CustomPrefs. If a property does exist in CustomPrefs, then the deserialized version of the value from CustomPrefs is used for that property. If the property does not exist in CustomPrefs, the default value is used for that property.

File: frontend/src/components/Layout/Images/__tests__/BlurhashCanvas.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (ferferga) [reply frontend/src/components/Layout/Images/__tests__/BlurhashCanvas.spec.ts:17]: @aweebs Yes please :)

File: frontend/src/utils/playback-profiles/helpers/codec-profiles.ts (hunk 3 above)
Comment (ferferga) [reply frontend/src/utils/playback-profiles/helpers/codec-profiles.ts:39]: @aweebs Future work, but I would rule modifying this file altogether in this PR, it's indeed introducing more type errors (even after applying the Vuetify 3.1.11 update to this branch).

In my discussions with @ThibaultNocchi, we thought the best approach for MediaCapabilities would be to phase out the "generate a device profile of all codecs at startup" approach and move to a "test for compatibility before starting the playback", based on the codec information of the selected media streams given by the server. We should test for video and audio independently and report as compatible all the web-based subtitlte formats (WebVTT) + ASS (thanks to JASSUB).

If a file is not compatible, we should report to server that we're compatible with H264 and AAC, which is a good default combo that all browsers supports. There's no point in checking for MP3, H263, etc... Maybe we can also test for H265 as Edge on Windows and Safari supports it.

That should be done exclusively in a new PR, as removing all this logic will break playback.

File: frontend/src/utils/external-config.ts (hunk 2 above)
Comment (ferferga) [reply frontend/src/utils/external-config.ts:50]: @aweebs Fair enough!

**Comment by ferferga** [comment]
@aweebs This has more type errors than current vite (this branch sits at 36 vs 20 at vite), even after the Vuetify upgrade.
Please, revisit this.

File: frontend/src/utils/playback-profiles/helpers/codec-profiles.ts (hunk 3 above)
Comment (aweebs) [reply frontend/src/utils/playback-profiles/helpers/codec-profiles.ts:39]: The added type errors are because of an outdated version of typescript. I realized I had been using 5.x.x since it was announced recently. Even upgrading to a later version of 4.x.x resolves these. I'll add the typescript update to this PR, it doesn't seem to impact anything other than our typing.

I would like to get rid of this type error though, the other cheating way to get rid of this is with an any cast
```
  if (isTizen()) {
    isTizenFhd = !(window as any).webapis.productinfo.isUdPanelSupported();
  }
```

I'd prefer to keep the above now that it won't cause additional type errors. Even if it's very verbose, we're pretty close to getting of all of them :D

**Comment by aweebs** [comment]
> @aweebs This has more type errors than current vite (this branch sits at 36 vs 20 at vite), even after the Vuetify upgrade. Please, revisit this.

Type Script upgrade was all that's needed 🎉
```
Found 6 errors in 5 files.

Errors  Files
     1  src/components/Layout/Carousel/Item/ItemsCarouselTitle.vue:15
     2  src/components/Layout/Carousel/Item/ItemsCarousel.vue:37
     1  src/plugins/remote/auth/index.ts:109
     1  src/utils/images.ts:198
     1  src/utils/items.ts:303
```

**Review by ferferga — APPROVED** [review]
